fetching/strat.py

```diff
@@ -29,8 +35,61 @@ def _test_file_filter(self, path: pathlib.Path) -> bool:
 
 
 class PyTestStrategy(ApplicationStrategy):
+    FUNCTION_PATTERN = re.compile(r"[\w\s]*test_[\w\s]*\([\w\s]*\)[\w\s]*:[\w\s]*")
+    IMPORTS = "from tracing import register, entrypoint\n"
+    REGISTER = "@register()\n"
+    ENTRYPOINT = "\n@entrypoint()\ndef main():\n  ...\n"
+    APPENDED_FILEPATH = "_decorators_appended.py"
+
+    def __init__(self, recurse_into_subdirs: bool = True):
+        super().__init__(recurse_into_subdirs)
+
+        self.decorator_appended_file_paths: list[pathlib.Path] = []
+
     def _apply(self, path: pathlib.Path) -> None:
-        pass
+        with path.open("r") as file:
+            lines = file.readlines()
+
+        skip_line = False
+        contains_pytest_test_function = False
+        for i, line in enumerate(lines):
+            if skip_line:
+                skip_line = False
+                continue
+            if PyTestStrategy.FUNCTION_PATTERN.fullmatch(line):
+                lines.insert(i, PyTestStrategy.REGISTER)
+                skip_line = True
+                contains_pytest_test_function = True
+
+        if contains_pytest_test_function:
+            lines.insert(0, PyTestStrategy.IMPORTS)
+            lines.append(PyTestStrategy.ENTRYPOINT)
+
+        file_path_with_appended_decorators = pathlib.Path(
+            str(path).replace(".py", PyTestStrategy.APPENDED_FILEPATH)
+        )
+        with file_path_with_appended_decorators.open("w") as file:
+            file.writelines(lines)
+
+        self.decorator_appended_file_paths.append(file_path_with_appended_decorators)
 
     def _test_file_filter(self, path: pathlib.Path) -> bool:
-        pass
+        if (
+            path.name.startswith("test_")
+            and path.name.endswith(".py")
+            and not path.name.endswith(PyTestStrategy.APPENDED_FILEPATH)
+        ):
+            return True
+
+        return path.name.endswith("_test.py")
+
+    def execute_decorator_appended_files(self):
```

Comment:
I am pretty sure that the execution is strategy-independent as it only executes the python file. Move this to the base class.

Reply:
If not, at least add the abstract function signature to the base class.

Reply:
The code is still there because it will be moved to another class that shall be instantiated from the method

Reply:
Something like 
```py
def apply(self, project_folder) -> TestExecutor:
    ... 
    return TestExecutor(files_with_decorators)
```

Reply:
I don't think creating a class for this functionality is necessary. Especially because the execution requires the file paths which are collected from the apply()-function.

Reply:
It should be necessary; each class should be responsible for precisely one thing, no more, no less. Currently, appending generated decorators and executing the resulting test files are part of `PyTestStrategy`, which is wrong from a clean code perspective.

Reply:
It depends on what you define as the responsibility of the class. If the goal of the strategy class is just the appending the decorators of the files, then that is okay. But from how I see it, it is more efficient to directly execute the generated files as we only care about the trace data files and not the generated test files. The extra abstraction is unnecessary as the following execution has to get the test file paths again and always requires the previous test file generation, while the generation is only needed for the execution and nothing else. The execution and strategy class would be coupled tightly which is pretty bad from a design perspective.

Reply:
> If the goal of the strategy class is just the appending the decorators of the files, then that is okay. 
Correct, that is its sole responsibility. It can later return a list of filepaths to generated files, avoiding having to search for the newly generated test paths

Reply:
Alright.